File: wlr/OGUR/OGUR/Paths/Path.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using OGUR.Collision;

namespace OGUR.Paths
{

    public class Path
    {
        public Point2 Finish = new Point2(0, 0);
        private List<Point2> m_steps = new List<Point2>();
        private Dictionary<Point2,Point2> m_stepLookup = new Dictionary<Point2,Point2>();
        private float m_totalWeight = 0;

        public Path(){}

        public Path Reset(Point2 start, Point2 finish)
        {
            m_steps.Clear();
            m_stepLookup.Clear();
            m_totalWeight = 0;
            moveIndex = -1;
            Finish.Copy(finish);
            Add(start);
            return this;
        }

        public Path Copy(Path source)
        {
            if (source != null)
            {
                m_stepLookup = StepLookup.Copy(source.m_stepLookup);
                m_steps = Walk.Copy(source.m_steps);
                m_totalWeight = source.m_totalWeight;
                Finish.Copy(source.Finish);
            }
            moveIndex = -1;
            return this;
        }

        public bool Add(Point2 step)
        {
            if(!m_stepLookup.ContainsKey(step))
            {
                m_stepLookup.Add(step, step);
                m_steps.Add(step);
                m_totalWeight += step.Weight;
                return true;
            }
            return false;
        }
        public float GetCost()
        {
            return m_totalWeight;
        }

        private int moveIndex = -1;
        public bool HasMoves()
        {
            return moveIndex < m_steps.Count();
        }
        public Point2 GetNextMove()
        {
            moveIndex++;
            if (moveIndex >= m_steps.Count())
            {
                return null;
            }
            if (m_steps.Count() == 0)
            {
                return null;
            }
            return m_steps.Count == 1 ? m_steps[0] : m_steps[moveIndex];
        }

        public bool IsDone()
        {
            return m_stepLookup.ContainsKey(Finish);
        }
        public Point2 GetLastStep()
        {
            return m_steps[m_steps.Count - 1];
        }

        public List<Point2> GetNeighbors()
        {
           return GetLastStep().GetNeighbors();
        }

        public int Length()
        {
            return m_steps.Count;
        }
    }
}

File: proj/src/spx/paths/Path.java
package spx.paths;

import spx.core.Point2;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Path {
    public final Point2 Finish = new Point2(0, 0);
    private List<Point2> _steps = new ArrayList<>();
    private HashMap<Point2, Point2> _stepLookup = new HashMap<>();
    private float _totalWeight = 0;

    public Path() {
    }

    public Path reset(Point2 start, Point2 finish) {
        _steps.clear();
        _stepLookup.clear();
        _totalWeight = 0;
        moveIndex = -1;
        Finish.copy(finish);
        add(start);
        return this;
    }

    public Path copy(Path source) {
        if (source != null) {
            _stepLookup = StepLookup.copy(source._stepLookup);
            _steps = Walk.copy(source._steps);
            _totalWeight = source._totalWeight;
            Finish.copy(source.Finish);
        }
        moveIndex = -1;
        return this;
    }

    public boolean add(Point2 step) {
        if (!_stepLookup.containsKey(step)) {
            _stepLookup.put(step, step);
            _steps.add(step);
            _totalWeight += step.Weight;
            return true;
        }
        return false;
    }

    public float getCost() {
        return _totalWeight;
    }

    private int moveIndex = -1;

    public boolean hasMoves() {
        return moveIndex < _steps.size();
    }

    public Point2 getNextMove() {
        moveIndex++;
        if (moveIndex >= _steps.size()) {
            return null;
        }
        if (_steps.size() == 0) {
            return null;
        }
        return _steps.size() == 1 ? _steps.get(0) : _steps.get(moveIndex);
    }

    public boolean isDone() {
        return _stepLookup.containsKey(Finish);
    }

    public Point2 getLastStep() {
        return _steps.get(_steps.size() - 1);
    }

    public List<Point2> getNeighbors() {
        return getLastStep().getNeighbors();
    }

    public int length() {
        return _steps.size();
    }
}
